File: extender/src/benchy_manager.c
#include "benchy_manager.h"
#include <string.h>

// Switch names for specific switch index values
static const char *SWITCH_NAMES[] = {
    "stbd",
    "port",
    "aft",
    "motor",
    "horn"
};

void benchy_create_config(BenchyMsg_t *msg, const Configuration_t *config) {
    msg->message_type = BENCHY_CONFIG;
    memcpy(&msg->payload.config, config, sizeof(Configuration_t));
}

void benchy_create_op(BenchyMsg_t *msg, const Operation_t *op) {
    msg->message_type = BENCHY_OP;
    memcpy(&msg->payload.op, op, sizeof(Operation_t));
}

#include <stdbool.h>

bool benchy_parse(BenchyMsg_t *msg, Benchy_t *data) {

     
    if (msg->message_type == BENCHY_CONFIG) {
        data->config.mode = msg->payload.config.mode;
        memcpy(data->config.mac_address, msg->payload.config.mac_address, 6);
        return true;
    } else if (msg->message_type == BENCHY_OP) {
        data->op.switch_value = msg->payload.op.switch_value;   
        memcpy(data->op.servo_values, msg->payload.op.servo_values, 4);
        return true;
    }

    return false; // Unknown message type
}


bool benchy_get_switch(uint8_t switch_value, int index) {
    return (switch_value & (1 << index)) != 0;
}

void benchy_set_switch(uint8_t *switch_value, int index, bool value) {
    if (value) {
        *switch_value |= (1 << index);
    } else {
        *switch_value &= ~(1 << index);
    }
}

const char* benchy_get_switch_name(int index) {
    if (index >= 0 && index < sizeof(SWITCH_NAMES) / sizeof(SWITCH_NAMES[0])) {
        return SWITCH_NAMES[index];
    }
    return "unknown";
}

void benchy_print_msg(const BenchyMsg_t *msg) {
    if (msg->message_type == BENCHY_CONFIG) {
        ESP_LOGI("benchy_manager", "Message Type: Configuration_t");
        ESP_LOGI("benchy_manager", "Mode: %d", msg->payload.config.mode);
        ESP_LOGI("benchy_manager", "MAC Address: %02X:%02X:%02X:%02X:%02X:%02X",
               msg->payload.config.mac_address[0], msg->payload.config.mac_address[1],
               msg->payload.config.mac_address[2], msg->payload.config.mac_address[3],
               msg->payload.config.mac_address[4], msg->payload.config.mac_address[5]);
    } else if (msg->message_type == BENCHY_OP) {
        ESP_LOGI("benchy_manager", "Message Type: Operation_t");
        ESP_LOGI("benchy_manager", "Switch Value: %d", msg->payload.op.switch_value);
        for (int i = 0; i < 8; ++i) {
            ESP_LOGI("benchy_manager", "%s: %s", benchy_get_switch_name(i), benchy_get_switch(msg->payload.op.switch_value, i) ? "true" : "false");
        }
        ESP_LOGI("benchy_manager", "Servo Values: %d, %d, %d, %d",
               msg->payload.op.servo_values[0], msg->payload.op.servo_values[1],
               msg->payload.op.servo_values[2], msg->payload.op.servo_values[3]);
    } else {
        ESP_LOGI("benchy_manager", "Unknown Message Type");
    }
}

void benchy_print(const Benchy_t *msg) {
        ESP_LOGI("benchy_manager", "MAC Address: %02X:%02X:%02X:%02X:%02X:%02X",
               msg->config.mac_address[0], msg->config.mac_address[1],
               msg->config.mac_address[2], msg->config.mac_address[3],
               msg->config.mac_address[4], msg->config.mac_address[5]);
        ESP_LOGI("benchy_manager", "Switch Value: %d", msg->op.switch_value);
        for (int i = 0; i < 8; ++i) {
            ESP_LOGI("benchy_manager", "%s: %s", benchy_get_switch_name(i), benchy_get_switch(msg->op.switch_value, i) ? "true" : "false");
        }
        ESP_LOGI("benchy_manager", "Servo Values: %d, %d, %d, %d",
               msg->op.servo_values[0], msg->op.servo_values[1],
               msg->op.servo_values[2], msg->op.servo_values[3]);
   
}


void benchy_init(Benchy_t *data) {
    Operation_t op = {
        .switch_value = 0x0f,
        .servo_values = {10, 20, 30, 40}
    };
    Configuration_t config = {
        .mode = MODE_UNIDIRECTIONAL,
        .mac_address = {1, 2, 3, 4, 5, 6}
    };
    data->op = op;
    data->config = config;
}

File: CTBenchy/app/src/main/java/com/cte/ctbenchy/BenchyManager.kt
package com.cte.ctbenchy
import android.util.Log
import java.nio.ByteBuffer
import java.nio.ByteOrder
class BenchyManager {

    // Define mode values
    companion object {
        const val TAG ="BenchManager"
        const val MODE_UNIDIRECTIONAL = 0x00.toByte()
        const val MODE_BIDIRECTIONAL = 0x01.toByte()
        const val MODE_PROGRAM = 0x02.toByte()
        val modeNames    = listOf("Uni-Directional", "Bi-Directional", "program")


        // Define servo channel indexes
        const val SERVO_CHANNEL_1 = 0
        const val SERVO_CHANNEL_2 = 1
        const val SERVO_CHANNEL_3 = 2
        const val SERVO_CHANNEL_4 = 3

        const val RUDDER = SERVO_CHANNEL_1
        const val THROTTLE = SERVO_CHANNEL_2

        // Define message types
        const val BENCHY_CONFIG = 0x10.toByte()
        const val BENCHY_OP = 0x20.toByte()

        // Define switch names for specific switch index values
        const val SWITCH_STBD = 0
        const val SWITCH_PORT = 1
        const val SWITCH_AFT = 2
        const val SWITCH_HORN = 3
        const val SWITCH_MOTOR = 4




        fun getRudder(data: Benchy): Int {
            return  ((data.operation.servoValues[RUDDER].toInt() and 0xFF) * 180 / 255) - 90 // Map 0..255 to -90..90
        }

        fun setRudder(data: Benchy, value: Int) {
           // rudder servo scalse -90 to 90 degrees but is limited in the screen slider to -60 to + 60
            var scaledValue = ((value + 90) * 255) / 180 // Map -90..90 to 0..255
            data.operation.servoValues[RUDDER] = (scaledValue and 0xff).toByte()
         }

        fun getThrottle(data: Benchy): Int {

            return if (data.config.mode == MODE_BIDIRECTIONAL) {
                ((data.operation.servoValues[THROTTLE].toInt() and 0xFF) * 180 / 255) - 90 // Map 0..255 to -90..90
            } else {
                ((data.operation.servoValues[THROTTLE].toInt() and 0xFF) * 180 / 255) // Map 0..255 to 0 .. 180
            }
        }

        fun setThrottle(data: Benchy, value: Int) {
            var scaledValue = if (data.config.mode == MODE_BIDIRECTIONAL) {
                ((value + 90) * 255) / 180 // Map -90..90 to 0..255
            } else {
                (value  * 255) / 180 // Map -90..90 to 0..255
            }
            data.operation.servoValues[THROTTLE] = (scaledValue and 0xff).toByte()
        }

        fun getSwitchValue(data: Benchy, index: Int): Boolean {
            return (data.operation.switchValue.toInt() and (1 shl index)) != 0
        }

        fun setSwitchValue(data: Benchy, index: Int, value: Boolean): Byte {
            require(index >= 0 && index < 8) { "Index must be between 0 and 7" }

            return if (value) {
                // Set the bit at index in switchValue
                (data.operation.switchValue.toInt() or (1 shl index)).toByte()
            } else {
                // Clear the bit at index in switchValue
                (data.operation.switchValue.toInt() and (1 shl index).inv()).toByte()
            }
        }

        fun getSwitchName(index: Int): String {
            return when (index) {
                SWITCH_STBD -> "stbd"
                SWITCH_PORT -> "port"
                SWITCH_AFT -> "aft"
                SWITCH_MOTOR -> "motor"
                SWITCH_HORN -> "horn"
                else -> "unknown"
            }
        }


        fun printBenchy(benchy: Benchy) {
            val config = benchy.config as? Configuration ?: return
            Log.i(TAG,"Mode: ${modeNames[config.mode.toInt()]}")
            Log.i(TAG,"MAC Address: ${config.macAddress.joinToString(":") { "%02X".format(it) }}")
            val op = benchy.operation as? Operation ?: return
            Log.i(TAG,"Switch Value: ${op.switchValue}")
            Log.i(TAG,"Throttle  ${BenchyManager.getThrottle(benchy)}")
            Log.i(TAG,"rudder  ${BenchyManager.getRudder(benchy)}")

        }
    }


    data class Configuration(
        var mode: Byte,
        val macAddress: ByteArray
    ) {
        init {
            require(macAddress.size == 6) { "MAC address must be 6 bytes long." }
        }

        fun toByteArray(): ByteArray {
            val buffer = ByteBuffer.allocate(7)
            buffer.order(ByteOrder.LITTLE_ENDIAN)
            buffer.put(mode)
            buffer.put(macAddress)
            return buffer.array()
        }

        companion object {
            fun fromByteArray(data: ByteArray): Configuration {
                val buffer = ByteBuffer.wrap(data)
                buffer.order(ByteOrder.LITTLE_ENDIAN)
                val mode = buffer.get()
                val macAddress = ByteArray(6)
                buffer.get(macAddress)
                return Configuration(mode, macAddress)
            }
        }
    }

    data class Operation(
        var switchValue: Byte,
        val servoValues: ByteArray
    ) {
        init {
            require(servoValues.size == 4) { "Servo values array must be 4 bytes long." }
        }

        fun toByteArray(): ByteArray {
            val buffer = ByteBuffer.allocate(5)
            buffer.order(ByteOrder.LITTLE_ENDIAN)
            buffer.put(switchValue)
            buffer.put(servoValues)
            return buffer.array()
        }

        companion object {
            fun fromByteArray(data: ByteArray): Operation {
                val buffer = ByteBuffer.wrap(data)
                buffer.order(ByteOrder.LITTLE_ENDIAN)
                val switchValue = buffer.get()
                val servoValues = ByteArray(4)
                buffer.get(servoValues)
                return Operation(switchValue, servoValues)
            }
        }
    }

    data class Benchy(
        val config: Configuration,
        val operation: Operation
    ) {
        fun toByteArray(): ByteArray {
            val configBytes = config.toByteArray()
            val operationBytes = operation.toByteArray()

            val buffer = ByteBuffer.allocate(configBytes.size + operationBytes.size)
            buffer.order(ByteOrder.LITTLE_ENDIAN)
            buffer.put(configBytes)
            buffer.put(operationBytes)
            return buffer.array()
        }

        companion object {
            fun fromByteArray(data: ByteArray): Benchy {
                val configData = data.copyOfRange(0, 7)
                val operationData = data.copyOfRange(7, 12)

                val config = Configuration.fromByteArray(configData)
                val operation = Operation.fromByteArray(operationData)

                return Benchy(config, operation)
            }
        }
    }



}

